tss-esapi/src/abstraction/mod.rs

```diff
@@ -6,3 +6,32 @@ pub mod cipher;
 pub mod ek;
 pub mod nv;
 pub mod transient;
+
+use crate::attributes::ObjectAttributesBuilder;
+use crate::utils::Tpm2BPublicBuilder;
+
+/// KeyCustomizaion allows to adjust how a key is going to be created
+pub trait KeyCustomization {
+    /// Alter the attributes used on key creation
+    fn attributes(&self, attributes_builder: ObjectAttributesBuilder) -> ObjectAttributesBuilder {
+        attributes_builder
+    }
+
+    /// Alter the key template used on key creation
+    fn template(&self, template_builder: Tpm2BPublicBuilder) -> Tpm2BPublicBuilder {
+        template_builder
+    }
+}
+
+// We set default value to Type parameters to function,
+// as a consequence, we can't just write:
+// ```
+// ak::create_ak(ctx, handle, Sha256, SignatureScheme, auth, None)
+// ```
+// as rust type checker will complain that it can't infer type for None.
+// To mitigate for that, provide a default implementation for (). This is
+// meant to be used like:
+// ```
+// ak::create_ak::<()>(ctx, handle, Sha256, SignatureScheme, auth, None)
+// ```
+impl KeyCustomization for () {}
```

Comment:
I am not so sure I like that the default or 'I do not care about customization' name is ```()``` couldn't it be called ```DefaultKeyCustomization``` or something like that?

Reply:
I'm not entirely happy about that solution either.
I went with `()` as it's easy to type in. My concern was to not introduce too much burden on the user, and thought `create_ak::<DefaultKeyCustomization>(_,_,_,_,_, None)` would undermine the usability.

Reply:
Here is another proposal. Relying on two methods instead. This avoids headaches with type inference.

Reply:
Another alternative would be:
https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=17fc37fc6d2ad309b0c52e4208bbf233

Reply:
Is it not possible to define it with a default implementation?
```pub fn foo<K: KeyCustomization = DefaultKeyCustomization>(.............).```

Reply:
No :( it has been deprecated :(
https://github.com/rust-lang/rust/issues/36887